docs/proposals/docker-registry-split.md

```diff
@@ -0,0 +1,141 @@
+# Move docker-registry to separate repository
+
+## Problem
+
+The `docker/distribution` has a lot of dependencies. Since `kubernetes` began to use docker/distribution
+it has become very difficult to update `docker/distribution` in the part where you need to combine the
+dependencies of these projects. There is no guarantee that the merge of the dependencies would not make
+regression in these projects.
+
+## Constraints and Assumptions
+
+The docker-registry is highly integrated in a common code-base so it is very difficult to describe
+all the difficulties of separation.
+
+## Investigation
+
+### etcd
+
+The docker-registry does not have its own `etcd`. All metadata are stored to the `origin` database using
+client API.
+
+### code-base
+
+The `origin` interacts with the docker-registry only through client library (no internal calls),
+but this cannot be said about docker-registry.
+
+1. docker-registry should use [clientsets](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/generating-clientset.md)
+of `origin`.
+2. docker-registry should use [clientsets](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/generating-clientset.md)
+of `kubernetes`.
+3. We need to create versioned client (clientsets?) for docker-registry as well.
+4. We need to split some utilities to reduce the number of dependencies to `origin`.
+5. We need to make common framework for tests. Currently docker-registry uses a common test utilities and clients.
```

Comment:
we should also start enabling the upstream docker-registry tests. we also need to enable the tests for different storage drivers to prove the storage works (remember GCS? :-)

i think this should be also mentioned here as part of the new infrastructure.